dipy/core/gradients.py

```diff
@@ -674,7 +674,87 @@ def round_bvals(bvals, bmag=None):
     return b.round() * (10 ** bmag)
 
 
-def unique_bvals(bvals, bmag=None, rbvals=False):
+def unique_bvals_tolerance(bvals, tol=20):
+    """ Gives the unique b-values of the data, within a tolerance gap
+
+    The b-values must be regrouped in clusters easily separated by a
+    distance greater than the tolerance gap. If all the b-values of a
+    cluster fit within the tolerance gap, the highest b-value is kept.
+
+    Parameters
+    ----------
+    bvals : ndarray
+        Array containing the b-values
+
+    tol : int
+        The tolerated gap between the b-values to extract
+        and the actual b-values.
+
+    Returns
+    ------
+    ubvals : ndarray
+        Array containing the unique b-values using the median value
+        for each cluster
+    """
+    b = np.unique(bvals)
+    ubvals = []
+    i = 0
+    lower_part = np.where(b <= b[i] + tol)[0]
+    upper_part = np.where(np.logical_and(b <= b[lower_part[-1]] + tol,
+                                         b > b[lower_part[-1]]))[0]
+    ubvals.append(b[lower_part[-1]])
+    if len(upper_part) != 0:
+        i = upper_part[-1] + 1
+    else:
+        i = lower_part[-1] + 1
+    while i != len(b):
+        lower_part = np.where(np.logical_and(b <= b[i] + tol,
+                                             b > b[i-1]))[0]
+        upper_part = np.where(np.logical_and(b <= b[lower_part[-1]] + tol,
+                                             b > b[lower_part[-1]]))[0]
+        ubvals.append(b[lower_part[-1]])
+        if len(upper_part) != 0:
+            i = upper_part[-1] + 1
+        else:
+            i = lower_part[-1] + 1
+
+    # Checking for overlap with get_bval_indices
+    for i in range(len(ubvals)-1):
+        indices_1 = get_bval_indices(bvals, ubvals[i], tol)
+        indices_2 = get_bval_indices(bvals, ubvals[i+1], tol)
+        if len(np.intersect1d(indices_1, indices_2)) != 0:
+            msg = '''There is overlap in clustering of b-values.
+            The tolerance factor might be too high.'''
+            warn(msg, UserWarning)
```

Comment:
is it ok to have an overlap of clustering? Why don't you raise an error?

Reply:
It could result in problems for the user, but I didn't know if it could still be used in some cases. I can raise an error.

Reply:
IMO: I think a warning is okay here!

Reply:
ok, thanks for the explanation @karanphil

Reply:
you can keep it